[MRG+1] Fix #10229: check_array should fail if array has strings

rtlee9 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #10229 


#### What does this implement/fix? Explain your changes.
Adds a deprecation warning if a non-object string-like array (defined as being a subdtype of `np.flexible`) is passed to `check_array` with `dtype="numeric"`. Arrays with object, boolean, and number dtypes are handled as before.


#### Any other comments?
The added deprecation warning indicates that non-object string-like arrays will be handled as object arrays are currently handled, i.e., attempted to be converted to `np.float64`. It seems intuitive to me to treat all string-like arrays (object + flexible) the same, but please let me know if you prefer any alternatives.

For reference: [numpy scalar dtypes](https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html)

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

rtlee9 committed:
Add deprecation warning to check_array for flexible array w/ dtype=numeric

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -476,6 +476,13 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+        # in the future np.flexible dtypes will be handled like object dtypes
+        if dtype_numeric and np.issubdtype(array.dtype, np.flexible):
+            warnings.warn(
+                "In the future, array with dtype {} will be handled as object,"
```
Plese follow the conventions of the message given in http://scikit-learn.org/dev/developers/contributing.html#deprecation.

Unless I miss something you should say explicitly in the message that it will cause an error (rather than handled as objects which probably does not mean much to the end user).

rtlee9 replied:
Here's an example where I think it would make sense to handle in the same way as an object array rather than raise an error:
let `X_str_num = [['1', '2'], ['3', '4']]` then `check_array(np.array(X_str_num, 'O'), 'numeric')` returns an array with dtype float64 but `check_array(X_str_num, 'numeric')` and `check_array(np.array(X_str_num, 'U1'), 'numeric')` return an array with dtype U1. 

This behavior for object arrays is expected base on the docstring, so we would just have to expand that description from object to object + flexible at the end of the deprecation cycle if we do want to treat flexible arrays the same as objects (e.g., attempt to cast to numeric).
> If "numeric", dtype is preserved unless array.dtype is object.

rtlee9 replied:
To align the deprecation warning message with the guidelines, I've added the version the change in behavior is to be expected to the warning in commit 4bcfc92. I'm not sure if we need to indicate the current version, since no changes have actually been made this version. For this reason I also haven't added a deprecation note to the docstring, but please let me know if you'd suggest otherwise.

lesteve commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -245,6 +245,22 @@ def test_check_array():
 
+    # deprecation warning if string-like array with dtype="numeric"
+    X_str = [['a', 'b'], ['c', 'd']]
+    assert_warns(DeprecationWarning, check_array, X_str, "numeric")
```
Please check the warning message through assert_warns_message.

rtlee9 replied:
Thanks @lesteve, fixed in commit 50cd194

rtlee9 committed:
Use assert_warns_message to test check_array deprecation warning

rtlee9 committed:
Add deprecation starting version to warning message

rtlee9 committed:
NumPy may convert string list to dtype U1 or S1 so need to check both in test

lesteve commented:
> Here's an example where I think it would make sense to handle in the same way as an object array rather than raise an error:
let `X_str_num = [['1', '2'], ['3', '4']]` then `check_array(np.array(X_str_num, 'O'), 'numeric')` returns an array with dtype float64 but `check_array(X_str_num, 'numeric')` and `check_array(np.array(X_str_num, 'U1'), 'numeric')` return an array with dtype U1. 

IMO this is all these weird string to float silent conversions that we want to disable in the future. I was not aware that numpy could convert strings to float64 like this:

```py
import numpy as np
np.array(['1', '222']).astype(np.float64)
```

I could not find a `casting` argument that allowed objects to float64 only if there were numbers in the array, please double-check. There may be an alternative approach.

To sum up what I think we should aim for in the future (after a 2 release deprecation cycle):
```py
check_array(np.array(['1', '2'], dtype='O'), dtype='numeric') # error
check_array(np.array(['1', '2'], dtype='S1'), dtype='numeric') # error
check_array(np.array(['1', '2'], dtype='U1'), dtype='numeric') # error
check_array(np.array([1, 2], dtype='O'), dtype='numeric') # no error
```

rtlee9 commented:
I couldn't find a `casting` argument in the documentation. I took another look at the code as well, and it looks like an array with dtype 'O' is set to be cast to `np.float64` in [this](https://github.com/scikit-learn/scikit-learn/blob/202b5321f1798c4980abf69ac8c0a0969f01a2ec/sklearn/utils/validation.py#L415) line, conditional only on being an object type and the dtype argument being numeric.

On the other hand there is a `warn_on_dtype` argument, which raises a DataConversionWarning if the dtype of the input array doesn't match the dtype of the returned array. Having this argument makes me feel less less weary of the automatic string -> numeric conversions, as the user has the option of being warned when this happens.

rtlee9 commented:
Took a look at your examples as well - it sounds like you're suggesting we look at the elements in a numpy array to determine what the underlying type is as opposed to relying on the np.array.dtype attribute.

I haven't found a good way of doing that without checking the type of potentially every element in an array yet; for example with `np.array([[1, 2, 3, '4']], dtype='O')` we wouldn't know that there's a string element in this array (i.e., in order to know whether to throw an error in `check_array`) without iterating through it, right?

I'd have to do some more digging if this is the route we want to go -- seems like something that should be handled by numpy ideally, but I'm not sure. Should this functionality be opened as a separate issue perhaps?

lesteve commented:
It seems like `.astype(casting='safe')` does what we want to do in the future. Maybe we can have something like:

```py
try:
    array = array.astype(dtype=np.float64, casting='safe')
except TypeError:
    warnings.warn('...') 
    array = array.astype(dtype=np.float64, casting='unsafe')
```

rtlee9 commented:
Sounds good with the exception that object arrays even with no strings cannot be cast to float with `casting='safe'`:
```python
np.array([1, 2], dtype='O').astype(np.float64, casting='safe')  # raises TypeError
```

lesteve commented:
Maybe object -> float64 conversion is a conversion that we want to disable in the future, not 100% sure. If we are fine with silently converting object to float64, we can have a special case in the code.

rtlee9 commented:
Agreed. In my opinion these silent conversions aren't completely unacceptable, especially given there's a `warn_on_dtype` argument to warn for this kind of thing.

I think this is a step in the right direction at least -- no more silently returning a string array when numeric is requested. Any additional suggestions for this PR?

jnothman reviewed (commented):
While I suspect this is being unnecessarily cautious (because the current behaviour is a bug, or an oversight), I'm okay with this fix-slowly strategy.

I think that the error message should be improved: "Arrays of strings will be interpreted as decimal numbers" rather than "will be handled as arrays with dtype object". I think we should (in the future?) outright reject arrays of `bytes` objects.

jnothman commented:
Btw I've briefly checked test and example build logs to ascertain that we're not currently raising this warning anywhere.

rtlee9 committed:
Make warning message more clear

lesteve commented:
I am looking at this with fresh eyes and I am wondering actually why we can not have an exception (without deprecation period) when `dtype.kind is in ('S', 'U')` (which covers arrays of bytes/str/unicode in Python 2/3). I would think that even if `check_array` lets through an array like `np.array(['1', '2'])`, an error will be raised when this array actually gets used.

Maybe we should have an error as well for `dtype.kind == 'V'`, in which case `np.issubdtype(..., np.flexible)` (as used in this PR at the time of writing) is a good way of doing it.

jnothman commented:
how about we start by writing a common test that says estimators should
reject decimal string input. If all estimators in scikit-learn already
raise an error for this case, I'm alright to do so in check_array.
Otherwise, I think a conservative approach like the present one is
sufficient for now.

rtlee9 commented:
These examples suggest at least some estimators don't raise an error for decimal strings
```python
from sklearn.linear_model import LogisticRegression
LogisticRegression().fit([['1', '1'], ['0', '1']], [0, 1])  # no error
LogisticRegression().fit(np.array([['1', '1'], ['0', '1']], dtype='O'), [0, 1])  # no error
LogisticRegression().fit(np.array([['1', '1'], ['0', '1']], dtype='U'), [0, 1])  # no error

from sklearn.svm import SVC
SVC().fit([['1', '1'], ['0', '1']], [0, 1])  # no error
SVC().fit(np.array([['1', '1'], ['0', '1']], dtype='O'), [0, 1])  # no error
SVC().fit(np.array([['1', '1'], ['0', '1']], dtype='U'), [0, 1])  # no error
```

I think the estimators call `check_X_y` from their `fit` method with `dtype=np.float64`, which calls `check_array` with the same dtype. `check_array` then attempts to convert to `np.foat64` and raises a ValueError if it can't. As it's being used now it seems like the logical place for raising a decimal string error would be in `check_array` so I'm not too surprised it isn't raised elsewhere in the estimator.

jnothman commented:
so estimators currently setting dtype explicitly convert decimal strings.
weird then that we should be inconsistent across estimators. I would rather
we didn't support strings at all, but I'd prefer to do so than to support
in some and not others.

rtlee9 commented:
It does look like the use of `check_X_y` / `check_array` at least is inconsistent across estimators; for example, [MLP](https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/neural_network/multilayer_perceptron.py#L1303) doesn't pass a dtype, while logistic regression, SVM, and GBR use float32 or float64. 

> I would rather we didn't support strings at all, but I'd prefer to do so than to support in some and not others.

^^ Makes sense to me.

rtlee9 commented:
Got it. So is there anything else I can do for this pull request? If I'm interpreting your comments correctly, we don't want to reject decimal strings in `check_array` quite yet.

jnothman commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -285,6 +285,32 @@ def test_check_array():
+    X_byte = [[b'a', b'b'], [b'c', b'd']]
+    assert_warns_message(
+        DeprecationWarning,
+        "arrays of strings will be interpreted as decimal numbers "
```
I still think we should aim to reject, rather than interpret, bytes

jnothman reviewed (commented):
But otherwise, yes, this is reasonable

rtlee9 committed:
Update check_array warning: reject byte arrays if numeric dtype

rtlee9 committed:
Update test to account for numpy conversion of string list to byte array

NumPy may convert string list to dtype U1 or S1 so need to accept
both deprecation warnings

lesteve commented:
Maybe I am missing something but this feels like a very partial solution to the original problem:
* this only applies to dtype='numeric'. A lot of the estimators specify `dtype` (hacky awk-based search seems to indicate ~40% of the check_array calls specify dtype outside tests).
* I am not really sure why we treat string arrays differently than byte arrays. I think we should strive to raise an error for both eventually.
* I really don't understand the warning ""Beginning in version 0.22, arrays of strings will be interpreted as decimal numbers if parameter 'dtype' is numeric". I think arrays of strings are already (and probably have been for some time) interpreted as decimal numbers.

jnothman commented:
I think atm with dtype='numeric' you would get an object array back, and
this would cause errors unless otherwise explicitly cast.

I object to byte arrays in python 3 being interpreted as strings because
python 3 makes a principled stand against doing so.

we are trying to maintain backwards compatibility for current uses of
dtype=float or similar, which currently eval decimal strings (and bytes:
yuck!). i don't really like this behaviour, but it's not especially
harmful, add it's what some users may expect of numpy. I'd be happy to
deprecate it, but for now this takes a more conservative approach.

Does that make some sense?

lesteve commented:
OK that makes sense, thanks a lot for the details ! Somehow in my head I wanted this PR to be about silent conversion of strings/bytes to numbers ... I'll try to review in more details to move this PR forward.

jnothman commented:
I'm happy to deprecate decimal number string support, but we need to do so in the dtype='float' case too

lesteve commented:
I have looked at this a bit more and I would be in favour of doing the following things:
* not making a difference between arrays of strings and arrays of bytes. I would do the same as numpy which can convert both fine to float64 (for better or worse). Making a difference between array of strings and array of bytes also introduces a slight discrepancy between `dtype='numeric'` and `dtype=np.float64`
* the warning should probably be a FutureWarning (rather than a DeprecatingWarning)
*  the warning needs an advice on how to get rid of the warning (basically it is strongly recommended to use .astype(np.float64) or .astype(np.float32)  before feeding it into scikit-learn)
* possibly adding a whats_new entry

jnothman commented:
Let's go with that plan. Sorry @rtlee9 for the back and forth. A what's new is a good idea because a few users reported this as a bug

rtlee9 committed:
No differentiation between arrays of strings and arrays of bytes

rtlee9 committed:
Raise FutureWarning instead of DeprecationWarning

rtlee9 committed:
Add recommendation to convert array to np.float64 to warning

rtlee9 committed:
Add whats_new bug fix entry

rtlee9 committed:
Merge branch 'master' into check_array

rtlee9 commented:
No worries -- these all sounded like good suggestions to me. I've added a few commits (one per item in @lesteve 's list) and merged in master to remove conflicts in the whats_new entry. Please let me know if you have any edits, especially around warning/whats_new messaging.

lesteve committed:
Tweak tests and whats_new entry

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -335,6 +335,12 @@ Feature Extraction
+Utils
+
+- Fixed a bug in :func:`utils.validation.check_array` to raise a ``FutureWarning``
+  indicating that arrays of type ``np.flexible`` will be interpreted as decimal
```
Just say strings

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -516,6 +516,15 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
+        # in the future np.flexible dtypes will be handled like object dtypes
+        if dtype_numeric and np.issubdtype(array.dtype, np.flexible):
+            warnings.warn(
+                "Beginning in version 0.22, arrays of strings will be "
```
The standard user will very likely have little idea what check_array is. This is the wording I am proposing, better suggestions more than welcome:
```py
                "Beginning in version 0.22, arrays of bytes/strings will be "
                "interpreted as decimal numbers if dtype='numeric'. "
                "It is recommended that you convert the array to "
                "a float dtype before using it in scikit-learn, "
                "for example by using "
                "your_array = your_array.astype(np.float64).",
```

lesteve commented:
I pushed a tweak in the whats_new entry and in tests. The wording of the warning can maybe be improved but otherwise LGTM.

jnothman commented:
I'd like to hope not many people see the warning. Thanks @rtlee9

rtlee9 commented:
Welcome, and thank you both for the reviews

amueller commented:
I'm confused as to what this will do in the future. I don't understand the message :-/

amueller commented:
This warning happens in feature selection when we try to transform the feature names btw.

amueller commented:
so the future behavior is to convert them to float? Can we change "interpret" to "convert"?

amueller commented:
Also we should be clearer what the current behavior is, which is passing through arbitrary strings if dtype='numeric'.
